subsys/bluetooth/mesh/access.c

```diff
@@ -2614,3 +2634,220 @@ uint8_t bt_mesh_comp_parse_page(struct net_buf_simple *buf)
+	}
+}
+
+static struct access_message *get_access_msg(const struct bt_mesh_model *model)
+{
+	struct access_message *msg = NULL;
+
+	for (int i = 0; i < ARRAY_SIZE(access_messages); i++) {
+		if (!access_messages[i].busy) {
+			msg = &access_messages[i];
+		}
+
+		if (access_messages[i].model == model) {
+			msg = &access_messages[i];
+			break;
+		}
+	}
+
+	if (!msg) {
+		LOG_DBG("No available access message context. Purge pending access message.");
+		msg = access_randomizer.pending_msg;
+	}
+
+	return msg;
+}
+
+static int push_msg_from_randomizer(struct access_message *msg)
+{
+	sys_snode_t *node;
+	struct access_message_chunk *chunk;
+	size_t len = msg->len;
+	int err;
+
+	NET_BUF_SIMPLE_DEFINE(buf, len);
+
+	SYS_SLIST_FOR_EACH_NODE(&msg->chunks, node) {
+		size_t tmp = MIN(CONFIG_BT_MESH_DELAYED_ACCESS_MSG_CHUNK_SIZE, len);
+
+		chunk = CONTAINER_OF(node, struct access_message_chunk, node);
+		memcpy(net_buf_simple_pull_mem(&buf, tmp), chunk->data, tmp);
+		len -= tmp;
+	}
+
+	err = push_msg_to_transport(msg->model, &msg->ctx, &buf, msg->cb, msg->cb_data);
+	if (err != -EBUSY && err != -ENOBUFS) {
+		release_msg_chunks(msg);
+		msg->busy = false;
+		if (err && msg->cb && msg->cb->end) {
+			msg->cb->end(err, msg->cb_data);
+		}
+	}
+
+	reschedule_randomizer();
+
+	return err;
+}
+
+static int push_msg_to_randomizer(const struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx,
+				  struct net_buf_simple *buf, const struct bt_mesh_send_cb *cb,
+				  void *cb_data)
+{
+	sys_snode_t *node;
+	struct access_message *msg;
+	struct access_message *purge_msg;
+	uint16_t random_delay;
+	int total_number = ROUND_UP(buf->size, CONFIG_BT_MESH_DELAYED_ACCESS_MSG_CHUNK_SIZE);
+	int allocated_number = 0;
+	size_t len = buf->len;
+	int err;
+
+	if (total_number > CONFIG_BT_MESH_DELAYED_ACCESS_MSG_CHUNK_COUNT) {
+		return -EINVAL;
+	}
+
+	msg = get_access_msg(model);
+	purge_msg = msg->busy ? msg : NULL;
+
+	do {
+		if (purge_msg) {
+			LOG_DBG("Purge access message from model %i on elem %i", model->rt->mod_idx,
+				model->rt->elem_idx);
+			err = push_msg_from_randomizer(purge_msg);
+			if (err == -EBUSY && err == -ENOBUFS) {
+				return -ENOMEM;
+			}
+		}
+
+		allocated_number += allocate_msg_chunks(msg, total_number - allocated_number);
+
+		if (total_number > allocated_number) {
+			LOG_DBG("Unable allocate %u chunks, allocated %u", total_number,
+				allocated_number);
+			purge_msg = access_randomizer.pending_msg;
+		}
+	} while (total_number > allocated_number);
+
+	SYS_SLIST_FOR_EACH_NODE(&msg->chunks, node) {
+		size_t tmp = MIN(CONFIG_BT_MESH_DELAYED_ACCESS_MSG_CHUNK_SIZE, len);
+
+		struct access_message_chunk *chunk =
+			CONTAINER_OF(node, struct access_message_chunk, node);
+
+		memcpy(chunk->data, net_buf_simple_pull_mem(buf, tmp), tmp);
+		len -= tmp;
+	}
+
+	bt_rand(&random_delay, sizeof(uint16_t));
+	random_delay = 20 + random_delay % (BT_MESH_ADDR_IS_UNICAST(ctx->recv_dst) ? 30 : 480);
```

Comment:
Why is there "20 +"? I think this part should be removed. Messages originating from access layer will result in unique ADV payloads.

Reply:
Each model can choose to ignore random delay. So, no problem.